pilot/pkg/proxy/envoy/v2/debounced_channel.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"sync"
+	"time"
+
+	"istio.io/istio/pilot/pkg/model"
+)
+
+// DebouncedEvent is used as entry type, which is enqueued in DebouncedChannel
+type DebouncedEvent = interface{}
+
+// DebouncedChannel is a channel which abstracts the debounce algorithm
+type DebouncedChannel struct {
+	// C channel, which can be used in selects. It delivers debounced events.
+	// The function emitted must be called. Otherwise processing is stopped.
+	C <-chan func(func(DebouncedEvent))
+	c chan func(func(DebouncedEvent))
+
+	pushCounter      uint32
+	debounceCounter  uint32
+	debounceAfterEnd time.Time
+	debounceMaxEnd   time.Time
+	lastEnqueueTime  time.Time
+	lastEmitTime     time.Time
+	debounceAfter    time.Duration
+	debounceMax      time.Duration
+	mutex            sync.Mutex
+
+	merge              func(accumulator DebouncedEvent, value DebouncedEvent) DebouncedEvent
```

Comment:
maybe a personal preference, but do we need to make this abstract/generic? We already know this is only used for the pushing debouncing, so we already know how merge works

Reply:
In my point of view, there are 2 reasons to keep this generic:
1. Testing with simple boolean values is easier
2. As far as I can judge, `DebouncedChannel` could be also a replacement for `PushQueue`

Reply:
How would debounce channel replace pushqueue? push queue does not debounce, it is just a simple queue that will do deduping. I guess we could go even more generic and make a single object that can handle both, but it seems more complicated to me.

For the bool part - I would much rather see the test actually testing what we use than have the test be completely different - the update req struct is pretty simple so I don't think it would make the tests much more complex, but would make the code much simpler

Reply:
`PushQueue` also merges events. I would expect that `DebounceQueue` with `debounceAfter` set to zero would behave the same way. 
It would have the additional benefit that it could be multiplexed with `select`. This seems to be a problem in `doSendPushes`. In this function, events from the `stopCh` will never be processed if the `queue.Dequeue()` blocks forever.

Reply:
This debounce holds a single event at a time , but pushqueue holds many items at once. Maybe it's possible, but I personally don't see how